### PR TITLE
Correct drag tools sync

### PR DESCRIPTION
**Pull request by polycone**
- Added synchronization for tools:
  - Attack
  - Build (buildings, pipes, wires, vents, etc.)
  - Cancel
  - Copy settings (doesn't work as expected right now, because it requires source building settings sync; settings can't be placed into the command payload because they are building-dependent)
  - Deconstruct
  - Dig
  - Disconnect
  - Disinfect
  - Empty pipe
  - Harvest
  - Debug (cell modifications and templates)
  - Mop
  - Priority
  - Sweep
  - Wrangle

- Added a new, relatively safe, approach to "patch" in-game data for some operations (see `GameContext`):
  ```csharp
  GameContext.Override(new SpecificContext(params), () => { contextDependentOperation(); });
  ```

- Game static context (like, `Something.Instance` static fields) can be overridden with the `GameStaticContext`:
  ```csharp
  GameStaticContext.Override();
  staticContextDependentOperations();
  GameStaticContext.Restore();
  ```
  It's designed to be used inside other contexts, that's why it doesn't have the "using" method (like, `GameContext.Override`).

- Added additional game state data that can be accessed via, well, `GameState`.

### CODE REVIEW

File: MultiplayerMod/Multiplayer/Commands/Tools/Build.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+using MultiplayerMod.Game.Building;
+using MultiplayerMod.Game.Events.Tools;
+using UnityEngine;
+
+namespace MultiplayerMod.Multiplayer.Commands.Tools;
+
+[Serializable]
+public class Build : IMultiplayerCommand {
+
+    private BuildEventArgs @event;
```
Comment (zuev93) [inline MultiplayerMod/Multiplayer/Commands/Tools/Build.cs:13]: What is a meaning of `@` in the var name?

**Review by zuev93 — APPROVED** [review]

File: MultiplayerMod/Multiplayer/Commands/Tools/Build.cs (hunk 1 above)
Comment (Ceron257) [reply MultiplayerMod/Multiplayer/Commands/Tools/Build.cs:13]: > The @ character prefixes a code element that the compiler is to interpret as an identifier rather than a C# keyword.

From [C# reference](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/verbatim). In this case the `event` keyword is used as a variable name.

File: MultiplayerMod/Game/Building/BuildValidation.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+using System.Linq;
+using System.Reflection;
+using HarmonyLib;
+
+namespace MultiplayerMod.Game.Building;
+
+[HarmonyPatch]
+public static class BuildValidation {
+
+    public static bool Enabled { get; set; } = true;
```
Comment (zuev93) [inline MultiplayerMod/Game/Building/BuildValidation.cs:11]: It is quite hard to follow (for me personally). 
What do you think calling it something like `PatchEnabled` (and invert value as I get it)

File: MultiplayerMod/Game/GameState.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿using HarmonyLib;
+using UnityEngine;
+
+namespace MultiplayerMod.Game;
+
+[HarmonyPatch]
+public static class GameState {
```
Comment (zuev93) [inline MultiplayerMod/Game/GameState.cs:7]: Name seems to be miss aligned with patched class and the class purpose as well.
Maybe `GamePatch`, `GameOnPrefabInitPatch` or `LastSelectedObjectState` might be better.

File: MultiplayerMod/Multiplayer/Commands/Tools/Build.cs (hunk 1 above)
Comment (zuev93) [reply MultiplayerMod/Multiplayer/Commands/Tools/Build.cs:13]: Thanks for the explanation.
But I still concerned and don't really see a strong reason to use keyword for variable names here (and above for `override`) and getting to deal with this mechanism to override it.
Here for instance we can use `args`, `buildEvent` or even `buildEventArgs`.

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs [hunk 4]
```diff
@@ -37,32 +36,39 @@ public class GameEventBindings {
             () => client.Send(new UpdateCursorPosition(client.Player, position))
         );
 
-        BindTools();
+        DragToolEvents.DragComplete += (sender, args) => {
+            // @formatter:off
+            switch (sender) {
```
Comment (zuev93) [inline MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:41]: Nit:
As I get not all sender types must be handled here, so it is making impossible to rely on handling all the cases and using default as a catch branch.

Would be great if we can handle all the cases and do not think about what to skip here.

File: MultiplayerMod/Multiplayer/Commands/Tools/Build.cs (hunk 1 above)
Comment (polycone) [reply MultiplayerMod/Multiplayer/Commands/Tools/Build.cs:13]: This thing has another causality, it's not like
> use a keyword for variable name

It's the name happens to be a keyword in C#, and that's good the language allows to use keywords as identifiers.
Anyway, I'll think once more, at first sight `event` is the most suitable name in the command domain.

File: MultiplayerMod/Game/Building/BuildValidation.cs (hunk 2 above)
Comment (polycone) [reply MultiplayerMod/Game/Building/BuildValidation.cs:11]: Don't assume it's a patch, yet it contains patch instructions. API should be readable for maintainable code, so it's `BuildValidation.Enabled = true | false` and literally means enable or disable validation during build process.
Nevertheless it will probably be replaced with game context operations later (it's still a draft PR, mostly for me to track changes).

File: MultiplayerMod/Game/GameState.cs (hunk 3 above)
Comment (polycone) [reply MultiplayerMod/Game/GameState.cs:7]: What are the concerns? 
It's current game state that happens to have the last selected object (and also contains some instructions on how to obtain it).

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs (hunk 4 above)
Comment (polycone) [reply MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:41]: > As I get not all sender types must be handled here

Yes, only ones we support, so what is the reason for handling all cases and use default? I mean, we simply don't care about other types and don't need to check for "unusual" values, because we support only these that are explicitly defined, nothing more.

That's why I didn't get
>Would be great if we can handle all the cases and do not think about what to skip here.

All the supported cases are handled, others - don't care, not an error, not an exception, just full ignorance🙂.

File: MultiplayerMod/Multiplayer/Commands/Tools/Build.cs (hunk 1 above)
Comment (polycone) [reply MultiplayerMod/Multiplayer/Commands/Tools/Build.cs]: Decided to change it to `arguments` since this data is actually command arguments🤔.

File: MultiplayerMod/Game/Building/BuildValidation.cs (hunk 2 above)
Comment (polycone) [reply MultiplayerMod/Game/Building/BuildValidation.cs]: Replaced with `DisableBuildingValidation` context.

File: MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿namespace MultiplayerMod.Game.Tools.Events;
+
+public static class BuildToolPriority {
+
+    public static PrioritySetting Get() {
```
Comment (zuev93) [inline MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs:5]: Nit:
Why not getter?

File: MultiplayerMod/Game/Tools/Events/DebugToolEvents.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            return;
+
+        var context = new DebugToolContext {
+            AffectCells = DebugPaintElementScreen.Instance.affectCells.isOn,
+            AffectBuildings = DebugPaintElementScreen.Instance.affectBuildings.isOn,
+            PreventFowReveal = DebugPaintElementScreen.Instance.set_prevent_fow_reveal,
+            AllowFowReveal = DebugPaintElementScreen.Instance.set_allow_fow_reveal
+        };
+
+        if (DebugPaintElementScreen.Instance.paintElement.isOn)
```
Comment (zuev93) [inline MultiplayerMod/Game/Tools/Events/DebugToolEvents.cs:37]: Nit:

Can be moved to constructor with ternary operator.

Also DebugPaintElementScreen.Instance might be extracted to a variable.

File: MultiplayerMod/Multiplayer/Commands/Tools/CopySettings.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+namespace MultiplayerMod.Multiplayer.Commands.Tools;
+
+[Serializable]
+public class CopySettings : IMultiplayerCommand {
+
+    private static readonly Core.Logging.Logger log = LoggerFactory.GetLogger<CopySettings>();
+
+    private CopySettingsEventArgs @event;
+
+    public CopySettings(CopySettingsEventArgs @event) {
```
Comment (zuev93) [inline MultiplayerMod/Multiplayer/Commands/Tools/CopySettings.cs:16]: Should it be `arguments` as in other places? :)

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs (hunk 4 above)
Comment (zuev93) [reply MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:41]: It is more about writing a code here.

If a new tool will be added our mod will skip it. Which might be fine since it'll be working. But the missing sync part wont noticed.

**Review by zuev93 — APPROVED** [review]

File: MultiplayerMod/Multiplayer/Commands/Tools/CopySettings.cs (hunk 7 above)
Comment (polycone) [reply MultiplayerMod/Multiplayer/Commands/Tools/CopySettings.cs:16]: Whoops.

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs (hunk 4 above)
Comment (polycone) [reply MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:41]: Not sure it's a problem and not sure it can be unnoticed when something isn't synced (especially a tool😂), but you want to make it extra fool-proof? 
Maybe consider this in the future and create an issue?
What do you think about implementation? It could be like an in-game popup effect "This tool isn't supported in multiplayer yet" or it could crash the game with "Sorry, no way this gonna work".

File: MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs (hunk 5 above)
Comment (polycone) [reply MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs:5]: Why getter lmao? 
My reasoning is the class is a property itself and this `Get` is the getter.
Now I have a thought it could be placed into `GameState` right?

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs (hunk 4 above)
Comment (zuev93) [reply MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:41]: Me neither, thats why I said it is `Nit`.
So if you don't see any small change here to make it either ways here - I'm fine with current impl.

File: MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs (hunk 5 above)
Comment (zuev93) [reply MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs:5]: Usually `GetSomething` in C# could be and should be replaced with getters as more default and native way.
Unless it is not a getter but something like computation task with result or whatever else.

Move to GameState seems good to me.

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs (hunk 4 above)
Comment (polycone) [reply MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:41]: Okay 🙂
I've created an issue https://github.com/zuev93/oni_multiplayer/issues/59 just not to forget about it.

File: MultiplayerMod/Game/Tools/Events/DebugToolEvents.cs (hunk 6 above)
Comment (polycone) [reply MultiplayerMod/Game/Tools/Events/DebugToolEvents.cs:37]: Nice catch, thanks!

File: MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs (hunk 5 above)
Comment (polycone) [reply MultiplayerMod/Game/Tools/Events/BuildToolPriority.cs]: Moved.

File: MultiplayerMod/Game/GameState.cs (hunk 3 above)
Comment (polycone) [reply MultiplayerMod/Game/GameState.cs:7]: Now it's more obvious, let's review in the future.
